settings: add a Sensitive option and use it for some settings

rafiss opened the pull request:
This new option causes the value of a string setting to be redacted in all outputs that are shown to the user. Users who have the MODIFYCLUSTERSETTING privilege can view the unredacted setting.

Virtual cluster setting overrides don't have their privilege checks changed for the write path. The only change is that when using SHOW commands, the values of sensitive settings will be redacted if the user does not have MODIFYCLUSTERSETTING.

See individual commits for details to make it easier to review.

fixes https://github.com/cockroachdb/cockroach/issues/115356

Release note (security update): Introduced the `server.redact_sensitive_settings.enabled` cluster
setting, which is false by default. If set to true, then the values of the following settings will be redacted
when accessed through SHOW commands or other introspection interfaces:
- server.oidc_authentication.client_id
- server.oidc_authentication.client_secret

In the future, any other sensitive cluster settings that are added will be redacted as well.
Users who have the MODIFYCLUSTERSETTING privilege can view the unredacted
settings.

Release note (cluster virtualization): If the `server.redact_sensitive_settings.enabled` cluster setting
is set to true, then the MANAGEVIRTUALCLUSTER privilege is required to view the values of the
per-virtual-cluster overrides for sensitive cluster settings.

rafiss committed:
settings: add a Sensitive option and use it for some settings

This new option causes the value of a string setting to be redacted
in all outputs that are shown to the user. Users who have the
MODIFYCLUSTERSETTING privilege can view the unredacted setting.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/115851)
<!-- Reviewable:end -->

rafiss committed:
sql,settings: introduce and use LookupForDisplay when appropriate

Instead of using LookupForLocalAccessByKey in SHOW and
when populating the crdb_internal table, we add a LookupForDisplay
function. This new function will cause the value of any sensitive
cluster setting to be redacted when being shown, unless the user has
privileges to see sensitive settings.

For the redaction to work correctly, some of the setting encode/decode
logic had to be updated too.

Release note: None

rafiss committed:
logictestccl: add a test for sensitive settings

Release note: None

rafiss committed:
settings: add ValueOrigin and DefaultString to Setting interface

Release note: None

rafiss committed:
settings: add cluster setting to control redaction of sensitive settings

Release note (security update): Introduced the `server.redact_sensitive_settings.enabled`
cluster setting, which is false by default. If set to true, then the values
of the following settings will be redacted when accessed through SHOW commands
or other introspection interfaces:
- server.oidc_authentication.client_id
- server.oidc_authentication.client_secret

In the future, any other sensitive cluster settings that are added will be
redacted as well. Users who have the MODIFYCLUSTERSETTING privilege can always
view the unredacted settings.

Release note (cluster virtualization): If the
`server.redact_sensitive_settings.enabled` cluster setting is set to true, then
the MANAGEVIRTUALCLUSTER privilege is required to view the values of the
per-virtual-cluster overrides for sensitive cluster settings.

dhartunian reviewed (commented):
OIDC changes LGTM. Thanks!

> Reviewed 1 of 4 files at r1, 1 of 7 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115851#-:-NnFUrbw7KQhuRCcPGIR:b-szjh9p)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

dt commented:
Can you expand the PR text a bit to explain the intended behavior around permissions and roles for operations involving overrides of settings of other virtual clusters by the system tenant? you appear to be going out of your way to make some of those check MODIFYCLUSTERSETTING  now, whereas they previously used the (more privileged?) VIEWCLUSTERMETADATA since they involved crossing virtual cluster boundaries.

rafiss commented:
> Can you expand the PR text a bit to explain the intended behavior around permissions and roles for operations involving overrides of settings of other virtual clusters by the system tenant? you appear to be going out of your way to make some of those check MODIFYCLUSTERSETTING now, whereas they previously used the (more privileged?) VIEWCLUSTERMETADATA since they involved crossing virtual cluster boundaries.

Sure -- this PR doesn't change any of the existing checks for VIEWCLUSTERMETADATA. The only change is when displaying settings, whether it's a virtual cluster override or not, so that the two sensitive settings are only shown unredacted if you have the MODIFYCLUSTERSETTING privilege.

Added to the PR text.

dt commented:
> Could you point to an example of the behavior you mentioned?

I think `ShowTenantClusterSetting` in sql/tenant_set is run _by a user in the system tenant_ to inspect the value of any overrides that are being applied to a secondary tenant. 

This code notably does _not_ check for `VIEWCLUSTERSETTINGS` today, but rather `VIEWCLUSTERMETADATA` (or `CanManageTenant` on the write path). 

I understand requiring a user -- including a user in a secondary tenant -- having `MODIFYCLUSTERSETTING ` in order to show the current value of a sensitive setting -- in their cluster -- but I'm not seeing how that also suggests the system tenant user should have `privilege.MODIFYCLUSTERSETTING` to see if an override is set, when they can _set_ the override without it (since that is off `CanManageTenant`).

rafiss commented:
I see what you mean now. in that case i will update the check in `ShowTenantClusterSetting` so that it only looks for the MANAGEVIRTUALCLUSTER privilege, since that is what is needed to set a virtual cluster setting override.

I'll add that to the PR text as well as a `cluster virtualization` release note.

dt commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -4976,17 +4979,29 @@ value if you rely on the HLC for accuracy.`,
+				// To view the per-virtual-cluster overrides for sensitive settings, the
+				// MANAGEVIRTUALCLUSTER privilege is required, since that is the
+				// privilege required to set the override in the first place.
+				hasManage, err := evalCtx.SessionAccessor.HasGlobalPrivilegeOrRoleOption(ctx, privilege.MANAGEVIRTUALCLUSTER)
```
It seems a little odd to me to _decode_ a value we're being passed into redaction -- we already had the value when we called the function after all.

I get that this is being done here since this gets called in the internal query, but it seems weird.

What instead of messing with this builtin, and the internal query in sql/tenant_settings.go, we just said that SHOWing any override -- for all overrides regardless of sensitivity -- requires manage tenant rather than viewclustermetadata? would that simplify this?

rafiss replied:
that would simplify things for sure. it makes sense to me too, since (unintentionally or not), `SHOW CLUSTER SETTINGS FOR VIRTUAL CLUSTER ...` already required `MANAGEVIRTUALCLUSTER`, but `SHOW CLUSTER SETTING <name> FOR VIRTUAL CLUSTER` did not.

i will make this change

dt commented on pkg/sql/tenant_settings.go:
```diff
@@ -177,6 +177,8 @@ func (p *planner) ShowTenantClusterSetting(
 	}
 
 	name := settings.SettingName(strings.ToLower(n.Name))
+	// NB: We use LookupForLocalAccess since the displayed value is computed in
```
Github won't let me comment on it, but I think the line just above this could just be changed from `CheckPrivilege(... VIEWCLUSTERMETADATA)` to `CanManageTenant(...)` and then we don't need any of the other edits in this func nor in the builtin?

dt replied:
@rafiss just wanted to check if you had a chance to consider this suggestion?

rafiss replied:
i had missed these comments since it wasn't marked as "changes requested." yes this idea sounds good to me. (though the `setting.InternalKey()`->`setting.Name()` and the usage of `decode_cluster_setting()` changes are both cleanups that i think we should keep.)

btw, not being able to comment on non-edited lines is one of my top gripes with Github reviews as compared to Reviewable.

rafiss reviewed (commented):
TFTR!

bors r+
